pkg/kubelet/pod_workers.go

```diff
@@ -981,26 +981,30 @@ func calculateEffectiveGracePeriod(status *podSyncStatus, pod *v1.Pod, options *
 	// enforce the restriction that a grace period can only decrease and track whatever our value is,
 	// then ensure a calculated value is passed down to lower levels
 	gracePeriod := status.gracePeriod
-	// this value is bedrock truth - the apiserver owns telling us this value calculated by apiserver
-	if override := pod.DeletionGracePeriodSeconds; override != nil {
-		if gracePeriod == 0 || *override < gracePeriod {
+
+	// defaults the grace period value to the pod's specified termination grace period seconds if present
+	// the kubelet provided no requested value (graceful termination?)
+	if override := pod.Spec.TerminationGracePeriodSeconds; override != nil {
+		if *override < gracePeriod {
 			gracePeriod = *override
 		}
 	}
+
+	// this value is top priority - the apiserver owns telling us this value calculated by apiserver
+	if override := pod.DeletionGracePeriodSeconds; override != nil {
+		gracePeriod = *override
+	}
+
 	// we allow other parts of the kubelet (namely eviction) to request this pod be terminated faster
 	if options != nil {
 		if override := options.PodTerminationGracePeriodSecondsOverride; override != nil {
-			if gracePeriod == 0 || *override < gracePeriod {
+			if *override < gracePeriod {
```

Comment:
This is incorrect, we must never allow a grace period of zero to be sent to CRI.

Reply:
Perhaps I spoke too soon - why exactly did you remove this check?

Reply:
> This is incorrect, we must never allow a grace period of zero to be sent to CRI.

We will not send a grace period of zero to CRI, because `gracePeriod` will be assigned the value of `minimumGracePeriodInSeconds` when `gracePeriod` is less than `minimumGracePeriodInSeconds`.

ref: https://github.com/kubernetes/kubernetes/blob/237d3dfda70d898e8b7641a295f5bdfecd527c2e/pkg/kubelet/kuberuntime/kuberuntime_container.go#L765-L768